Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -1315,6 +1317,64 @@ def visit_ExprNode(self, node):
         # With statements are never inside expressions.
         return node
 
+class ComprehensionScopeTransform(CythonTransform, SkipDeclarations):
+    """
+    Moves expressions out of list/set/etc comprehensions so they have the correct scope
+
+    Effectively
+        cmp = [ a for a in some_expression ]
+    is changed to
+        tmp = some_expression
+        cmp = [ a for a in tmp ]
+
+    (with some complications to keep optimizations (e.g. if "some_expression" is a range
+    where possible)
+    """
+    def visit_ComprehensionNode(self, node):
+        itseq = node.loop.iterator.sequence
+        # there's no value in substituting these very simple expressions
+        # with a temp (but NameNode must be so that it's evaluated in
+        # the right scope)
+        if isinstance(itseq, ExprNodes.ConstNode):
+            return node
+        original_node = node
+
+        new_itseq = ComprehensionResultRefNode(itseq)
+        node.loop.iterator.sequence = new_itseq
+        node = ComprehensionEvalWithTempExprNode(new_itseq, node)
+
+        if (isinstance(itseq, ExprNodes.SimpleCallNode) and
+            len(itseq.args) <= 3 # range with 3 arguments is the
```

Comment:
Please always keep two spaces before the `#` to visually separate comments from the code they refer to.
https://www.python.org/dev/peps/pep-0008/#inline-comments
Also, a multi-line inline comment seems self-contradicting. :)